### PR TITLE
Benjaabdala/api connection

### PR DESCRIPTION
**Pull request by benjaglobant**
In this PR I have implemented the service call to fetch all the characters to be shown in a recycler view (will be implemented in the next PR). Also I have tested the model and viewmodel used to achieve the API request.
Gif showing the UI:
servicePR
Image showing the status OK of the API response:
image
Image showing the passed tests:
image
Image showing the coverage:
image

### CODE REVIEW

File: app/src/main/res/layout/fragment_all_characters.xml [hunk 1]
```diff
@@ -1,14 +1,43 @@
...
+    <TextView
+        android:id="@+id/fragment_all_characters_character_id"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:textSize="@dimen/text_size_character_id"
+        android:textColor="@android:color/white"
+        android:textAlignment="center"
+        app:layout_constraintTop_toTopOf="parent"/>
+
+    <TextView
```
Comment (hernanhorquin96) [inline app/src/main/res/layout/fragment_all_characters.xml:34]: You can add styles to these textviews

File: app/src/main/java/com/globant/marvelmvvm/data/repository/MarvelRepository.kt [hunk 2]
```diff
@@ -0,0 +1,30 @@
+package com.globant.marvelmvvm.data.repository
+
+import com.globant.marvelmvvm.data.entity.Character
+import com.globant.marvelmvvm.data.mapper.MarvelMapper
+import com.globant.marvelmvvm.data.service.api.MarvelApi
+import com.globant.marvelmvvm.data.service.requestgenerator.MarvelRequestGenerator
+import com.globant.marvelmvvm.util.Constants.Companion.NOT_FOUND
+import com.globant.marvelmvvm.util.Result
+
+class MarvelRepository {
```
Comment (MuroSar) [inline app/src/main/java/com/globant/marvelmvvm/data/repository/MarvelRepository.kt]: this should be MarvelService not repository

File: app/src/test/java/com/globant/marvelmvvm/allcharactersTest/AllCharactersViewModelTest.kt [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        whenever(validResult.data).thenReturn(charactersResponse)
+        runBlocking {
+            viewModel.fetchAllCharacters().join()
+        }
+        verify(mockedModel).getAllCharacters()
+        assertEquals(responseList[0].status, liveDataUnderTest.observedValues[0]?.peekContent()?.status)
+        assertEquals(responseList[1].status, liveDataUnderTest.observedValues[1]?.peekContent()?.status)
+        assertEquals(responseList[1].data, liveDataUnderTest.observedValues[1]?.peekContent()?.data)
+    }
+
```
Comment (MuroSar) [inline app/src/test/java/com/globant/marvelmvvm/allcharactersTest/AllCharactersViewModelTest.kt:92]: remove this extra line pls

**Review by MuroSar — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/globant/marvelmvvm/viewmodel/AllCharactersViewModel.kt [hunk 4]
```diff
@@ -1,7 +1,42 @@
...
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
 
-class AllCharactersViewModel: ViewModel(), AllCharactersContract.ViewModel {
+class AllCharactersViewModel(private val model: AllCharactersContract.Model) : ViewModel(), AllCharactersContract.ViewModel {
+
+    var mutableMainState: MutableLiveData<Event<Data<List<Character>>>> = MutableLiveData()
+    override val mainState: LiveData<Event<Data<List<Character>>>>
+        get() = mutableMainState
```
Comment (MarcosCGlobant) [inline app/src/main/java/com/globant/marvelmvvm/viewmodel/AllCharactersViewModel.kt]: you can use a simple assignment here: 
override val mainState: LiveData<Event<Data<List<Character>>>> = mutableMainState

File: app/src/main/java/com/globant/marvelmvvm/data/repository/MarvelRepository.kt (hunk 2 above)
Comment (benjaglobant) [reply app/src/main/java/com/globant/marvelmvvm/data/repository/MarvelRepository.kt]: Done!

File: app/src/main/res/layout/fragment_all_characters.xml (hunk 1 above)
Comment (hernanhorquin96) [inline app/src/main/res/layout/fragment_all_characters.xml:34]: You can add styles in these textviews

File: app/src/test/java/com/globant/marvelmvvm/allcharactersTest/AllCharactersViewModelTest.kt (hunk 3 above)
Comment (benjaglobant) [reply app/src/test/java/com/globant/marvelmvvm/allcharactersTest/AllCharactersViewModelTest.kt:92]: Done!

File: app/src/test/java/com/globant/marvelmvvm/allcharactersTest/AllCharactersViewModelTest.kt [hunk 5]
```diff
@@ -0,0 +1,111 @@
+package com.globant.marvelmvvm.allcharactersTest
+
+import androidx.arch.core.executor.testing.InstantTaskExecutorRule
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.Observer
+import com.globant.marvelmvvm.contract.AllCharactersContract
+import com.globant.marvelmvvm.data.entity.Character
+import com.globant.marvelmvvm.util.*
```
Comment (mauroLonghi) [inline app/src/test/java/com/globant/marvelmvvm/allcharactersTest/AllCharactersViewModelTest.kt]: pls not use wildcards

File: app/src/test/java/com/globant/marvelmvvm/allcharactersTest/AllCharactersViewModelTest.kt [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        responseList = listOf(
+            Data(status = Status.LOADING, data = null, error = null),
+            Data(status = Status.RESPONSE_SUCCESS, data = charactersResponse, error = null))
+        whenever(mockedModel.getAllCharacters()).thenReturn(validResult)
+        whenever(validResult.data).thenReturn(charactersResponse)
+        runBlocking {
+            viewModel.fetchAllCharacters().join()
+        }
+        verify(mockedModel).getAllCharacters()
+        assertEquals(responseList[0].status, liveDataUnderTest.observedValues[0]?.peekContent()?.status)
```
Comment (mauroLonghi) [inline app/src/test/java/com/globant/marvelmvvm/allcharactersTest/AllCharactersViewModelTest.kt]: This it is not really a hardcode, but you added in the componion, two values with the same value...so you can use it

File: app/src/main/java/com/globant/marvelmvvm/viewmodel/AllCharactersViewModel.kt (hunk 4 above)
Comment (benjaglobant) [reply app/src/main/java/com/globant/marvelmvvm/viewmodel/AllCharactersViewModel.kt]: Done!

File: app/src/main/res/layout/fragment_all_characters.xml (hunk 1 above)
Comment (benjaglobant) [reply app/src/main/res/layout/fragment_all_characters.xml:34]: Hernan this textView is temporary (for show that the API is giving me the correct data). In the next PR I will delete this and implement a recycler view to show the data correctly. Thanks anyway!

File: app/src/test/java/com/globant/marvelmvvm/allcharactersTest/AllCharactersViewModelTest.kt (hunk 5 above)
Comment (benjaglobant) [reply app/src/test/java/com/globant/marvelmvvm/allcharactersTest/AllCharactersViewModelTest.kt]: Done!

File: app/src/test/java/com/globant/marvelmvvm/allcharactersTest/AllCharactersViewModelTest.kt (hunk 6 above)
Comment (benjaglobant) [reply app/src/test/java/com/globant/marvelmvvm/allcharactersTest/AllCharactersViewModelTest.kt]: Yes, I forgot put them here. Thanks Mau!

**Review by lucianoeli — APPROVED** [review]

**Review by MarcosCGlobant — APPROVED** [review]

File: app/src/main/java/com/globant/marvelmvvm/util/Constants.kt [hunk 7]
```diff
@@ -0,0 +1,9 @@
+package com.globant.marvelmvvm.util
+
+class Constants {
+    companion object{
+        const val EMPTY_STRING = ""
```
Comment (mauroLonghi) [inline app/src/main/java/com/globant/marvelmvvm/util/Constants.kt]: The correct use of an a Contants class is to define constants into the objet...but if you put it into the current componion obj it lose sense. So I recommend to define a Object class with this values like :
object MIS_CONSTANTES_CHETAS {
     const val EMPTY_STRING = ""
     const val EMPTY_CHAR = ' '
}

Comment (benjaglobant) [reply app/src/main/java/com/globant/marvelmvvm/util/Constants.kt]: You are right, I should not use that name but I will use the approach jaja, thanks!

**Review by mauroLonghi — APPROVED** [review]

File: app/src/test/java/com/globant/marvelmvvm/allcharactersTest/AllCharactersViewModelTest.kt (hunk 3 above)
Comment (SartiMau) [reply app/src/test/java/com/globant/marvelmvvm/allcharactersTest/AllCharactersViewModelTest.kt:92]: this still here

**Review by SartiMau — APPROVED** [review]

**Review by hernanhorquin96 — APPROVED** [review]

**Pull request merged**